File: Day7/Part1.py
testValues = []
operandsList = []
target = 0

with open("C:\\Users\\J\\source\\repos\\aoc-2024\\Day7\\daySevenInput.txt") as file:
    for line in file:
        result = line.split(':')[0].strip()
        operands = line.split(':')[1].strip().split(' ')
        testValues.append(int(result))
        operandsList.append([int(x) for x in operands])

def method(list: list, runningTotal, results: set, index):
    if index >= len(list) - 1:
        if index == len(list) - 1:
            results.add(runningTotal + list[index])
            results.add(int(f"{runningTotal}{list[index]}"))
            results.add(runningTotal * list[index] if runningTotal > 0 else list[index])
        else:
            results.add(runningTotal)
        return
    
    if runningTotal > target:
        return
    if index == 0:
        method(list, int(f"{list[index]}{list[index+1]}"), results, index + 2)
    else:
        method(list, int(f"{runningTotal}{list[index]}"), results, index + 1)
    method(list, runningTotal + list[index], results, index + 1)
    method(list, runningTotal * list[index] if runningTotal > 0 else list[index], results, index + 1)

def hasCombination(index):
    possibleResults = set()
    method(operandsList[index], 0, possibleResults, 0)
    return testValues[index] in possibleResults

total = 0

for i in range(0, len(testValues)):
    target = testValues[i]
    hasComb = hasCombination(i)
    total += testValues[i] if hasComb else 0

print("TOTAL:", total)

# 10 19
# => [10 + 19], [10 * 19]

# 81 40 27
# => [81 + 40 + 27], [81 * 40 + 27], [81 + 40 * 27], [81 * 40 * 27],
#    [81 + 27 + 40], [81 * 27 + 40], [81 + 27 * 40], [81 * 27 * 40],
# => [40 + 81 + 27], [40 * 81 + 27], [40 + 81 * 27], [40 * 81 * 27],
#    [40 + 27 + 81], [40 * 27 + 81], [40 + 27 * 81], [40 * 27 * 81],
# => [27 + 40 + 81], [27 * 40 + 81], [27 + 40 * 81], [27 * 40 * 81],
#    [27 + 81 + 40], [27 * 81 + 40], [27 + 81 * 40], [27 * 81 * 40]

File: Day7/Part1.java
package Day7;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Stream;

public class Part1 {

    public String FILENAME = "C:\\Users\\J\\source\\repos\\aoc-2024\\Day7\\daySevenInput.txt";
    HashMap<Long, List<Long>> map = new HashMap<>();

    public Part1() {
        try (BufferedReader br = new BufferedReader(new FileReader(FILENAME))) {
            String line;
            while ((line = br.readLine()) != null) {
                Long res = Long.parseLong(line.split(":")[0].strip());
                List<Long> ints = Stream.of(line.split(":")[1].strip().split(" ")).map(x -> Long.parseLong(x)).toList();
                map.put(res, ints);
            }
        } catch (IOException exception) {
            System.err.println(exception);
        }
    }

    public void method(List<Long> values, Long runningTotal, List<Long> results, int index) {
        if (index == values.size() - 1) {
            results.add(runningTotal + values.get(index));
            results.add(runningTotal == 0 ? values.get(index) : runningTotal * values.get(index));
            return;
        } 

        method(values,
            runningTotal + values.get(index),
            results,
            index + 1);
        method(values,
            runningTotal == 0 ? values.get(index) : runningTotal * values.get(index),
            results,
            index + 1);
    }

    public static void main(String[] args) {
        Part1 p1 = new Part1();

        List<Long> result = new ArrayList<>();

        Long total = 0L;
        for (Long value: p1.map.keySet()) {
            p1.method(p1.map.get(value), 0L, result, 0);
            if (result.contains(value)) {
                total += value;
            }
        }

        System.out.println(total);
    }
}
